v2/internal/reconcilers/arm/arm_client_cache.go

```diff
@@ -242,3 +259,38 @@ func (c *ARMClientCache) getSecret(ctx context.Context, namespace string, secret
 func matchSecretData(old, new map[string][]byte) bool {
 	return reflect.DeepEqual(old, new)
 }
+
+// TODO: Not sure if its a right thing to put this here or create a separate package for identity/auth and put workloadIdentityCredential there?
+// The below bit is from https://github.com/Azure/azure-sdk-for-go/issues/15615#issuecomment-1211012677
+type workloadIdentityCredential struct {
+	assertion     string
+	tokenFilePath string
+	cred          *azidentity.ClientAssertionCredential
+	lastRead      time.Time
+}
+
+func newWorkloadIdentityCredential(tenantID, clientID, file string) (*workloadIdentityCredential, error) {
+	w := &workloadIdentityCredential{tokenFilePath: file}
+	cred, err := azidentity.NewClientAssertionCredential(tenantID, clientID, w.getAssertion, nil)
+	if err != nil {
+		return nil, err
+	}
+	w.cred = cred
+	return w, nil
+}
+
+func (w *workloadIdentityCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (azcore.AccessToken, error) {
+	return w.cred.GetToken(ctx, opts)
+}
+
+func (w *workloadIdentityCredential) getAssertion(context.Context) (string, error) {
+	if now := time.Now(); w.lastRead.Add(3600 * time.Second).Before(now) {
```

Comment:
3600 is too long I think? The way this works (AFAIK) is that it re-reads the token periodically. We need to pick a time which is significantly faster than the underlying token expiry? The example written by @chlowell had 5m here but possibly that was just a dummy value.

Can you confirm with him what we should be setting this value to? I can't actually tell if/when the `AssertionCallback` is called from a quick read of their code. I see it registered but not used in the `GetToken` call. Apologies if you've already figured that out and 3600 is in fact correct.

Either way, once we have an answer maybe leave a comment here (or by the `3600` const I mentioned above) that explains why it's 3600 (or whatever the other value is).

Reply:
The 5 minute assertion cache was the feature team's recommendation when we first designed this last year. @aramase, is that still the best approach?

Reply:
The default token expiry configured with Azure Workload Identity is `1h` and the max allowed value is `24h`. The `kubelet` process renews the token at 80% expiry (so every 40 mins). With those numbers, every 30 min would work fine.